File: pybfm/bfm/obj.py
from .libbfm import lib, ffi
from .material import Material
from .mesh import Mesh
from .rule import Rule
from .state import default_state

class CObj:
	def __init__(self, c_obj, mesh: Mesh, material: Material, rule: Rule):
		self.c_obj = c_obj

		self.mesh = mesh
		self.material = material
		self.rule = rule

class Obj(CObj):
	def __init__(self, mesh: Mesh, material: Material, rule: Rule):
		c_obj = ffi.new("bfm_obj_t*")
		assert not lib.bfm_obj_create(c_obj, default_state, mesh.c_mesh, material.c_material, rule.c_rule)

		super().__init__(c_obj, mesh, material, rule)

	def __del__(self):
		assert not lib.bfm_obj_destroy(self.c_obj)

File: libbfm/src/obj.c
#include <bfm/obj.h>

int bfm_obj_create(bfm_obj_t* obj, bfm_state_t* state, bfm_mesh_t* mesh, bfm_material_t* material, bfm_rule_t* rule) {
	obj->state = state;

	obj->mesh = mesh;
	obj->material = material;
	obj->rule = rule;

	return 0;
}

int bfm_obj_destroy(bfm_obj_t* obj) {
	(void) obj;
	return 0;
}
